File: contribution-pipeline/scheduler/backend-py/src/scheduler/result_core.py
"""
applyResult 核心逻辑 — 与 result-core.ts 1:1

applyResult(body, conn=None):
  conn 缺省时自己开连接 + 事务; 有 conn 时复用 (业务层"业务写入+推进"原子用)
"""
from __future__ import annotations
import json
import math
from typing import Any

import asyncpg

from .db import get_pool
from .output_validator import validate_node_output
from .router import compute_next_step
from .types import StepConfig

STEP_OUTPUT_CAP_BYTES = 64 * 1024


class ResultError(Exception):
    def __init__(self, http_status: int, code: str, message: str) -> None:
        super().__init__(message)
        self.http_status = http_status
        self.code = code


def _sanitize_step_output(output: Any) -> dict[str, Any]:
    if output is None or not isinstance(output, dict):
        raise ResultError(400, "BAD_OUTPUT", "step output must be a plain object")
    serialized = json.dumps(output)
    if len(serialized.encode()) > STEP_OUTPUT_CAP_BYTES:
        raise ResultError(400, "OUTPUT_TOO_LARGE", f"step output > {STEP_OUTPUT_CAP_BYTES} bytes ({len(serialized)} actual)")
    return output


def _build_isolated_envelope(
    prev: dict[str, Any],
    step_key: str,
    output: dict[str, Any],
) -> dict[str, Any]:
    prev_outputs = prev.get("outputs") or {}
    new_outputs: dict[str, Any] = {k: v for k, v in prev_outputs.items() if k != step_key}
    new_outputs[step_key] = output
    return {"payload": prev.get("payload", {}), "outputs": new_outputs, "tags": prev.get("tags", {})}


async def apply_result(body: dict[str, Any], conn: asyncpg.Connection | None = None) -> dict[str, Any]:
    if conn is not None:
        return await _apply_result_in(body, conn)
    pool = get_pool()
    async with pool.acquire() as c:
        async with c.transaction():
            return await _apply_result_in(body, c)


async def _apply_result_in(body: dict[str, Any], conn: asyncpg.Connection) -> dict[str, Any]:
    run_id = body["runId"]
    ob = await conn.fetchrow("SELECT * FROM outbox WHERE run_id = $1", run_id)
    if not ob:
        raise ResultError(404, "UNKNOWN_RUN", "run not found")

    ob = dict(ob)
    if ob["status"] in ("done", "failed"):
        return {"ok": True, "applied": False}

    import datetime
    if ob["status"] == "leased" and ob.get("expected_by"):
        exp = ob["expected_by"]
        if isinstance(exp, str):
            exp = datetime.datetime.fromisoformat(exp)
        if exp.replace(tzinfo=datetime.timezone.utc) < datetime.datetime.now(datetime.timezone.utc):
            raise ResultError(409, "LEASE_EXPIRED", "租约已过期，请重新领取任务")

    effective_body = dict(body)

    if body.get("status") == "success" and body.get("output") is not None:
        lookup = await conn.fetchrow(
            """SELECT COALESCE(s->>'nodeVersion','1.0') AS node_version,
                      nd.outputs_schema, nd.outputs_validation
               FROM items i
               JOIN pipeline_versions pv ON pv.id = i.pipeline_version_id
               JOIN LATERAL jsonb_array_elements(pv.steps) s ON TRUE
               LEFT JOIN node_definitions nd
                 ON nd.key = $1 AND nd.version = COALESCE(s->>'nodeVersion','1.0')
               WHERE i.id = $2 AND s->>'key' = $3
               LIMIT 1""",
            ob["node_key"], ob["item_id"], ob["step_key"],
        )
        if lookup and lookup["outputs_schema"]:
            schema = json.loads(lookup["outputs_schema"]) if isinstance(lookup["outputs_schema"], str) else lookup["outputs_schema"]
            validation_mode = lookup["outputs_validation"] or "strict"
            v = validate_node_output(
                node_key=ob["node_key"],
                node_version=lookup["node_version"],
                outputs_schema=schema,
                outputs_validation=validation_mode,
                output=body["output"],
            )
            if v:
                sample = "; ".join(f"{e.path}: {e.message}" for e in v.errors[:3])
                msg = f"节点 {ob['node_key']}@{lookup['node_version']} 输出不符 outputsSchema: {sample}"
                if validation_mode == "strict":
                    effective_body = {**body, "status": "failed", "output": None, "error": {"code": "OUTPUT_SCHEMA_VIOLATION", "message": msg, "retryable": False}}
                else:
                    import logging
                    logging.getLogger(__name__).warning(f"[output-validator:warn] run={run_id} {msg}")

    outcome = "success" if effective_body.get("status") == "success" else "failed"
    output_json = json.dumps(effective_body["output"]) if effective_body.get("output") else None
    error_json = json.dumps(effective_body["error"]) if effective_body.get("error") else None

    await conn.execute(
        """INSERT INTO attempts
           (tenant_id, run_id, item_id, task_id, step_key, node_key, attempt, outcome,
            output, error, worker_id, started_at, finished_at)
           VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,NOW())""",
        ob["tenant_id"], ob["run_id"], ob["item_id"], ob["task_id"],
        ob["step_key"], ob["node_key"], ob["attempt"], outcome,
        output_json, error_json, ob.get("leased_by"), ob.get("leased_at"),
    )

    if effective_body.get("status") == "success":
        return await _apply_success(effective_body, conn, ob)
    return await _apply_failed(effective_body, conn, ob)


async def _apply_success(body: dict[str, Any], conn: asyncpg.Connection, ob: dict[str, Any]) -> dict[str, Any]:
    it = dict(await conn.fetchrow("SELECT * FROM items WHERE id = $1", ob["item_id"]))
    pipe_row = await conn.fetchrow("SELECT steps FROM pipeline_versions WHERE id = $1", it["pipeline_version_id"])
    steps_raw = json.loads(pipe_row["steps"]) if isinstance(pipe_row["steps"], str) else pipe_row["steps"]
    pipeline = [StepConfig(**s) for s in steps_raw]
    loop_counts: dict[str, int] = json.loads(it["loop_counts"]) if isinstance(it.get("loop_counts"), str) else (it.get("loop_counts") or {})

    route = compute_next_step(pipeline, ob["step_key"], body.get("output") or {}, body.get("nextHint"), loop_counts)
    safe_output = _sanitize_step_output(body.get("output") or {})
    envelope_raw = json.loads(it["envelope"]) if isinstance(it["envelope"], str) else it["envelope"]
    new_envelope = _build_isolated_envelope(envelope_raw, ob["step_key"], safe_output)
    new_loop_counts = dict(loop_counts)
    if route.loop_increment:
        k = route.loop_increment["stepKey"]
        new_loop_counts[k] = new_loop_counts.get(k, 0) + route.loop_increment["value"]

    await conn.execute(
        "UPDATE items SET envelope=$1, current_step=$2, loop_counts=$3, updated_at=NOW() WHERE id=$4",
        json.dumps(new_envelope), route.next_step_key, json.dumps(new_loop_counts), ob["item_id"],
    )
    await conn.execute("UPDATE outbox SET status='done', updated_at=NOW() WHERE run_id=$1", ob["run_id"])

    if route.next_step_key not in ("done", "stuck"):
        next_step = next((s for s in pipeline if s.key == route.next_step_key), None)
        if next_step:
            await conn.execute(
                "INSERT INTO outbox (tenant_id, item_id, task_id, step_key, node_key, status, attempt) VALUES ($1,$2,$3,$4,$5,'pending',1)",
                ob["tenant_id"], ob["item_id"], ob["task_id"], route.next_step_key, next_step.node_key,
            )

    return {"ok": True, "applied": True, "nextStep": route.next_step_key}


async def _apply_failed(body: dict[str, Any], conn: asyncpg.Connection, ob: dict[str, Any]) -> dict[str, Any]:
    retryable = (body.get("error") or {}).get("retryable", True)
    node = await conn.fetchrow("SELECT * FROM node_definitions WHERE key=$1 ORDER BY version DESC LIMIT 1", ob["node_key"])
    pipe_row = await conn.fetchrow(
        "SELECT pv.steps FROM items i JOIN pipeline_versions pv ON pv.id=i.pipeline_version_id WHERE i.id=$1", ob["item_id"]
    )
    steps_raw = json.loads(pipe_row["steps"]) if isinstance(pipe_row["steps"], str) else pipe_row["steps"]
    step_policy = next((s.get("policy") for s in steps_raw if s.get("key") == ob["step_key"]), None)
    max_attempts = (step_policy or {}).get("maxAttempts") or (node["default_max_attempts"] if node else 3) or 3
    base_backoff = (step_policy or {}).get("baseBackoffMs") or 1000

    await conn.execute("UPDATE outbox SET status='failed', updated_at=NOW() WHERE run_id=$1", ob["run_id"])

    if retryable and ob["attempt"] < max_attempts:
        import datetime
        delay_ms = round(base_backoff * math.pow(2, ob["attempt"] - 1))
        scheduled_at = datetime.datetime.now(datetime.timezone.utc) + datetime.timedelta(milliseconds=delay_ms)
        await conn.execute(
            "INSERT INTO outbox (tenant_id, item_id, task_id, step_key, node_key, status, attempt, scheduled_at) VALUES ($1,$2,$3,$4,$5,'pending',$6,$7)",
            ob["tenant_id"], ob["item_id"], ob["task_id"], ob["step_key"], ob["node_key"], ob["attempt"] + 1, scheduled_at,
        )
        return {"ok": True, "applied": True, "retryAt": scheduled_at.isoformat()}

    await conn.execute("UPDATE items SET current_step='stuck', updated_at=NOW() WHERE id=$1", ob["item_id"])
    return {"ok": True, "applied": True, "dlq": True}

File: contribution-pipeline/scheduler/backend/src/result-core.ts
// 提交执行结果的核心逻辑
//
// 设计:
//  - 路由层 (api.ts) 调 applyResult(body) → 自己开 sql.begin
//  - 业务层 (business.ts) 想"业务写入 + 调度推进"原子时, 自己 sql.begin 后传 tx 进来
//  - 错误统一通过 ResultError 抛出, 调用方按 httpStatus/code 翻译

import { sql, type SQL } from "./db.ts";
import { computeNextStep } from "./router.ts";
import { validateNodeOutput, type ValidationMode } from "./output-validator.ts";
import type { Envelope, ItemRow, NodeDefinition, OutboxRow, Pipeline } from "./types.ts";

/**
 * 节点输出隔离: 一个 step 提交的 output 只能写自己的 outputs[stepKey] slot,
 * 不能影响 payload (原始数据) / tags (元数据) / outputs[其他 stepKey]。
 *
 * 双层防御:
 *   1. 应用层 (本函数): 显式逐字段构造 newEnvelope, 不 spread 用户 output 到 payload/tags
 *   2. DB 层 (1715000170000_envelope_isolation.sql trigger): payload/tags UPDATE 一律拒
 *
 * Output 形态校验:
 *   - 必须是 plain object (非 null/array/primitive); 否则 BAD_OUTPUT 400
 *   - 大小硬上限 64 KiB JSON (per-step 容量, 防个别节点输出膨胀拖死全 pipeline)
 */
const STEP_OUTPUT_CAP_BYTES = 64 * 1024;

function sanitizeStepOutput(output: unknown): Record<string, unknown> {
  if (output == null || typeof output !== "object" || Array.isArray(output)) {
    throw new ResultError(400, "BAD_OUTPUT", "step output must be a plain object");
  }
  const json = JSON.stringify(output);
  if (json.length > STEP_OUTPUT_CAP_BYTES) {
    throw new ResultError(
      400, "OUTPUT_TOO_LARGE",
      `step output > ${STEP_OUTPUT_CAP_BYTES} bytes (${json.length} actual)`,
    );
  }
  return output as Record<string, unknown>;
}

function buildIsolatedEnvelope(
  prev: Envelope,
  stepKey: string,
  output: Record<string, unknown>,
): Envelope {
  // 显式逐字段构造, 拒绝用户 output 越权写到 payload/tags/其他 step 的 slot
  const prevOutputs = prev.outputs ?? {};
  const newOutputs: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(prevOutputs)) {
    if (k !== stepKey) newOutputs[k] = v; // 其他 step 的 slot 原样保留, 不被本次 output 影响
  }
  newOutputs[stepKey] = output; // 本 step slot, 唯一允许写的位置
  return {
    payload: prev.payload,        // 原始数据不可变 (DB trigger 兜底)
    outputs: newOutputs,
    tags: prev.tags,              // tags 不变 (DB trigger 兜底)
  };
}

export interface ApplyResultBody {
  runId: string;
  status: "success" | "failed";
  output?: Record<string, unknown>;
  error?: { code: string; message: string; retryable?: boolean };
  nextHint?: string;
}

export type ApplyResultOutput =
  | { ok: true; applied: false }
  | { ok: true; applied: true; nextStep: string }
  | { ok: true; applied: true; retryAt: string }
  | { ok: true; applied: true; dlq: true };

export class ResultError extends Error {
  constructor(
    public httpStatus: number,
    public code: string,
    message: string,
  ) {
    super(message);
    this.name = "ResultError";
  }
}

// 主入口: tx 缺省时自己开事务, 提供时复用调用方事务
export async function applyResult(
  body: ApplyResultBody,
  tx?: SQL,
): Promise<ApplyResultOutput> {
  const runInTx = (t: SQL) => applyResultIn(body, t);
  if (tx) return runInTx(tx);
  return sql.begin((t) => runInTx(t as unknown as SQL)) as Promise<ApplyResultOutput>;
}

async function applyResultIn(body: ApplyResultBody, tx: SQL): Promise<ApplyResultOutput> {
  const obs = await tx<OutboxRow[]>`SELECT * FROM outbox WHERE run_id = ${body.runId}`;
  const ob = obs[0];
  if (!ob) throw new ResultError(404, "UNKNOWN_RUN", "run not found");

  if (ob.status === "done" || ob.status === "failed") {
    return { ok: true, applied: false };
  }

  if (ob.status === "leased" && ob.expected_by && new Date(ob.expected_by) < new Date()) {
    throw new ResultError(409, "LEASE_EXPIRED", "租约已过期，请重新领取任务");
  }

  // outputsSchema 强校验 (06-node-design §4.4):
  //   business 说 success 但输出不符契约 → strict 翻转成 failed (retryable=false), warn 仅 console
  //   翻转必须在写 attempts 之前, 让 attempts.outcome 反映"实际生效结果"
  let effectiveBody = body;
  if (body.status === "success" && body.output != null) {
    const lookup = await tx<{
      node_version: string;
      outputs_schema: Record<string, unknown> | null;
      outputs_validation: ValidationMode | null;
    }[]>`
      SELECT
        COALESCE(s->>'nodeVersion', '1.0') AS node_version,
        nd.outputs_schema,
        nd.outputs_validation
      FROM items i
      JOIN pipeline_versions pv ON pv.id = i.pipeline_version_id
      JOIN LATERAL jsonb_array_elements(pv.steps) s ON TRUE
      LEFT JOIN node_definitions nd
        ON nd.key = ${ob.node_key}
       AND nd.version = COALESCE(s->>'nodeVersion', '1.0')
      WHERE i.id = ${ob.item_id} AND s->>'key' = ${ob.step_key}
      LIMIT 1
    `;
    const row = lookup[0];
    if (row?.outputs_schema) {
      const violation = validateNodeOutput({
        nodeKey: ob.node_key,
        nodeVersion: row.node_version,
        outputsSchema: row.outputs_schema,
        outputsValidation: row.outputs_validation,
        output: body.output,
      });
      if (violation) {
        const mode: ValidationMode = row.outputs_validation ?? "strict";
        const sample = violation.errors.slice(0, 3).map((e) => `${e.path}: ${e.message}`).join("; ");
        const msg = `节点 ${ob.node_key}@${row.node_version} 输出不符 outputsSchema: ${sample}${violation.errors.length > 3 ? " ..." : ""}`;
        if (mode === "strict") {
          // 翻转 success → failed retryable=false. attempts 会记 outcome=failed + error=OUTPUT_SCHEMA_VIOLATION,
          // outbox 走 applyFailed 路径直接进 DLQ (重试无意义, driver 行为不变就还是不符 schema)
          effectiveBody = {
            ...body,
            status: "failed",
            output: undefined,
            error: { code: "OUTPUT_SCHEMA_VIOLATION", message: msg, retryable: false },
          };
        } else {
          // warn: 不翻转, 但留下记录. console 写 (audit 异步写跨 tx 不挡主流程, 这里足够)
          console.warn(`[output-validator:warn] run=${body.runId} ${msg}`);
        }
      }
    }
  }

  // 1. 写 attempts (派生 tenant_id 来自 outbox 行)
  await tx`
    INSERT INTO attempts (
      tenant_id, run_id, item_id, task_id, step_key, node_key, attempt, outcome,
      output, error, worker_id, started_at, finished_at
    ) VALUES (
      ${ob.tenant_id}, ${ob.run_id}, ${ob.item_id}, ${ob.task_id}, ${ob.step_key}, ${ob.node_key},
      ${ob.attempt}, ${effectiveBody.status === "success" ? "success" : "failed"},
      ${effectiveBody.output ? sql.json(effectiveBody.output) : null},
      ${effectiveBody.error ? sql.json(effectiveBody.error) : null},
      ${ob.leased_by}, ${ob.leased_at}, NOW()
    )
  `;

  if (effectiveBody.status === "success") {
    return await applySuccess(effectiveBody, tx, ob);
  }
  return await applyFailed(effectiveBody, tx, ob);
}

async function applySuccess(
  body: ApplyResultBody,
  tx: SQL,
  ob: OutboxRow,
): Promise<ApplyResultOutput> {
  const its = await tx<ItemRow[]>`SELECT * FROM items WHERE id = ${ob.item_id}`;
  const it = its[0];
  // item 钉的 pipeline 版本: save 后路由仍按原版本走
  const pipe = await tx<{ steps: Pipeline }[]>`
    SELECT steps FROM pipeline_versions WHERE id = ${it.pipeline_version_id}
  `;
  const route = computeNextStep(
    pipe[0].steps,
    ob.step_key,
    body.output ?? {},
    body.nextHint,
    it.loop_counts ?? {},
  );

  const safeOutput = sanitizeStepOutput(body.output ?? {});
  const newEnvelope = buildIsolatedEnvelope(
    {
      payload: it.envelope.payload ?? {},
      outputs: it.envelope.outputs ?? {},
      tags: it.envelope.tags ?? {},
    },
    ob.step_key,
    safeOutput,
  );
  const newLoopCounts = route.loopIncrement
    ? {
        ...(it.loop_counts ?? {}),
        [route.loopIncrement.stepKey]:
          (it.loop_counts?.[route.loopIncrement.stepKey] ?? 0) + route.loopIncrement.value,
      }
    : it.loop_counts;

  await tx`
    UPDATE items SET
      envelope = ${sql.json(newEnvelope)},
      current_step = ${route.nextStepKey},
      loop_counts = ${sql.json(newLoopCounts ?? {})},
      updated_at = NOW()
    WHERE id = ${ob.item_id}
  `;
  await tx`UPDATE outbox SET status='done', updated_at=NOW() WHERE run_id=${ob.run_id}`;

  if (route.nextStepKey !== "done" && route.nextStepKey !== "stuck") {
    const nextStep = pipe[0].steps.find((s) => s.key === route.nextStepKey)!;
    await tx`
      INSERT INTO outbox (tenant_id, item_id, task_id, step_key, node_key, status, attempt)
      VALUES (${ob.tenant_id}, ${ob.item_id}, ${ob.task_id}, ${route.nextStepKey}, ${nextStep.nodeKey}, 'pending', 1)
    `;
  }

  return { ok: true, applied: true, nextStep: route.nextStepKey };
}

async function applyFailed(
  body: ApplyResultBody,
  tx: SQL,
  ob: OutboxRow,
): Promise<ApplyResultOutput> {
  const retryable = body.error?.retryable !== false;
  const node = await tx<NodeDefinition[]>`
    SELECT * FROM node_definitions WHERE key = ${ob.node_key} ORDER BY version DESC LIMIT 1
  `;
  // policy 也走 item 钉的版本
  const pipe = await tx<{ steps: Pipeline }[]>`
    SELECT pv.steps FROM items i
    JOIN pipeline_versions pv ON pv.id = i.pipeline_version_id
    WHERE i.id = ${ob.item_id}
  `;
  const policy = pipe[0]?.steps.find((s) => s.key === ob.step_key)?.policy;
  const maxAttempts = policy?.maxAttempts ?? node[0]?.default_max_attempts ?? 3;
  const baseBackoff = policy?.baseBackoffMs ?? 1000;

  await tx`UPDATE outbox SET status='failed', updated_at=NOW() WHERE run_id=${ob.run_id}`;

  if (retryable && ob.attempt < maxAttempts) {
    const delayMs = Math.round(baseBackoff * Math.pow(2, ob.attempt - 1));
    const scheduledAt = new Date(Date.now() + delayMs);
    await tx`
      INSERT INTO outbox (
        tenant_id, item_id, task_id, step_key, node_key, status, attempt, scheduled_at
      ) VALUES (
        ${ob.tenant_id}, ${ob.item_id}, ${ob.task_id}, ${ob.step_key}, ${ob.node_key},
        'pending', ${ob.attempt + 1}, ${scheduledAt}
      )
    `;
    return { ok: true, applied: true, retryAt: scheduledAt.toISOString() };
  }

  await tx`UPDATE items SET current_step='stuck', updated_at=NOW() WHERE id=${ob.item_id}`;
  return { ok: true, applied: true, dlq: true };
}
